source/extensions/filters/http/cache/cache_insert_queue.cc

```diff
@@ -0,0 +1,194 @@
+  } else {
+    insert_context_->insertTrailers(
+        trailers, [this](bool cache_success) { onFragmentComplete(cache_success, true, 0); });
+    fragment_in_flight_ = true;
+  }
+}
+
+void CacheInsertQueue::onFragmentComplete(bool cache_success, bool end_stream, size_t sz) {
+  // If the cache implementation is asynchronous, this may be called from whatever
+  // thread that cache implementation runs on. Therefore, we post it to the
+  // dispatcher to be certain any callbacks and updates are called on the filter's
+  // thread (and therefore we don't have to mutex-guard anything).
+  dispatcher_.post([this, cache_success, end_stream, sz]() {
+    fragment_in_flight_ = false;
+    if (aborting_) {
+      // Parent filter was destroyed, so we can quit this operation.
+      self_ownership_.reset();
+      return;
+    }
+    ASSERT(queue_size_bytes_ >= sz, "queue can't be emptied by more than its size");
+    queue_size_bytes_ -= sz;
+    if (watermarked_ && queue_size_bytes_ <= low_watermark_bytes_) {
+      if (encoder_callbacks_.has_value()) {
+        encoder_callbacks_.value().get().onEncoderFilterBelowWriteBufferLowWatermark();
+      }
+      watermarked_ = false;
+    }
+    if (!cache_success) {
+      // canceled by cache; unwatermark if necessary, inform the filter if
+      // it's still around, and delete the queue.
+      if (watermarked_ && encoder_callbacks_.has_value()) {
+        encoder_callbacks_.value().get().onEncoderFilterBelowWriteBufferLowWatermark();
+      }
+      fragments_.clear();
+      self_ownership_.reset();
+      abort_callback_();
+      return;
+    }
+    if (end_stream) {
+      ASSERT(fragments_.empty(), "ending a stream with the queue not empty is a bug");
+      ASSERT(!watermarked_, "being over the high watermark when the queue is empty makes no sense");
+      self_ownership_.reset();
+      return;
+    }
+    if (!fragments_.empty()) {
+      // If there's more in the queue, push the next fragment to the cache.
+      auto fragment = std::move(fragments_.front());
+      fragments_.pop_front();
+      fragment_in_flight_ = true;
+      fragment->send(*insert_context_, [this](bool cache_success, bool end_stream, size_t sz) {
+        onFragmentComplete(cache_success, end_stream, sz);
+      });
+    }
+  });
+}
+
+void CacheInsertQueue::setSelfOwned(std::unique_ptr<CacheInsertQueue> self) {
+  // Disable all the callbacks, they're going to have nowhere to go.
+  abort_callback_ = []() {};
+  encoder_callbacks_.reset();
+  if (fragments_.empty() && !fragment_in_flight_) {
+    // If the queue is already empty we can just let it be destroyed immediately.
+    return;
+  }
+  if (!end_stream_queued_) {
+    // If the queue can't be completed we can abort early but we need to wait for
+    // any callback-in-flight to complete before destroying the queue.
+    aborting_ = true;
+  }
+  self_ownership_ = std::move(self);
+}
+
+CacheInsertQueue::~CacheInsertQueue() { insert_context_->onDestroy(); }
```

Comment:
maybe this should fire a bunch of invariant asserts on the various bools it tracks e.g. no watermarking, etc. as done by other code in this area

Reply:
Added a couple of those, which required changing some related behaviors (it was okay to destroy with watermarked_=true and/or queue not empty, so long as the low watermark event had been sent. Now clearing the flag happens every time the event is sent, and the queue is cleared when abort occurs, so the requested assertions are now correct.)